### ISSUE TITLE
Questions about using fmodel' weights as model's weight

### ISSUE THREAD

**Issue by xieshuqin** [facebookresearch/higher#42]
Thanks for the awesome work of higher! 

I am trying to implement [this paper](https://arxiv.org/pdf/2001.08735.pdf). The goal here is to jointly learn a set of hyperparameters as well as a base_model such that base_model trained on the given hyperparameters can perform well on the validation set. 

A snippet code looks like this:
```
import higher

base_model = Model()
ft_model = FTModel()  # a network to predict hyperparameters
base_opt = torch.optim.SGD(base_model.parameters(), lr=...)
ft_opt = torch.optim.SGD(ft_model.parameters(), lr=...)

for x_train, y_train, x_test, y_test in loader:
    with higher.innerloop_ctx(base_model, base_opt) as (fmodel, diffopt):
        hyper = ft_model(x_train)  # predict hyperparameters

        # update base_model given the hyperparameters
        train_loss = fmodel(x_train, y_train, hyper) 
        diffopt.step(train_loss) 

        test_loss = fmodel(x_test, y_test)
    # use test_loss to update ft_model
    test_loss.backward()
    ft_opt.step() 
```

My questions are:
1. Is it possible to directly assign `fmodel`'s weight to `base_model`? This is the update rule for `base_model` in the paper. 
2. Does `diffopt.step(train_loss)` also computes gradients w.r.t. `hyper`? We want the gradient of `hyper` to only come from `test_loss`'s gradient w.r.t. `base_model`'s initial weights.

**Comment by xieshuqin** [comment]
After looking at the source code, I found `diffopt.step(train_loss)` only computes gradients w.r.t. parameters at `diffopt`, so it won't compute gradients w.r.t. `hyper`. 

I am still not clear how to do fast weight assignment. Appreciate for any hints.

**Comment by egrefen** [comment]
> Is it possible to directly assign fmodel's weight to base_model? This is the update rule for base_model in the paper.

Have you tried simply using `state_dict`/`load_state_dict` methods? See [the pattern in this comment](https://github.com/facebookresearch/higher/issues/10#issuecomment-574695558) of #10 for reference. 
Let me know how you get on and if this works in your case (or doesn't), please.

> Does diffopt.step(train_loss) also computes gradients w.r.t. hyper? We want the gradient of hyper to only come from test_loss's gradient w.r.t. base_model's initial weights.

It shouldn't. `diffopt` branches from `base_opt` which only has a view on `base_model.parameters()`. When you call `test_loss = fmodel(x_test, y_test)` you compute the meta-loss as a function of the (implicit) updated weights of `fmodel`, but also creates an implicit backward graph spanning back to the original weights (with `hyper` as a node). Therefore, when you call `test_loss.backward()`, `hyper` receives gradient, which is then applied in `ft_opt.step()`.

> After looking at the source code, I found diffopt.step(train_loss) only computes gradients w.r.t. parameters at diffopt, so it won't compute gradients w.r.t. hyper.

That's right.

> I am still not clear how to do fast weight assignment. Appreciate for any hints.

See first point in this reply.

In the meantime, I'm closing this issue, but please re-open if you are still having trouble with the weight setting (or better yet, open a new issue with a minimal repro, stack trace, and any other comments).
